File: ios/NativeMonitoring/MotionModule.swift
import Foundation
import CoreMotion

/// Section 21/22/23 — mirrors android/.../MotionModule.kt's feature math so
/// both platforms report the same fields to the JS layer.
///
/// NOT BUILD-VERIFIED: written without Xcode/macOS access. Needs to be
/// added to the Xcode target and compiled on a Mac before use — see the
/// bridging-header note in MotionModule.m.
@objc(MotionModule)
class MotionModule: RCTEventEmitter {

  private let motionManager = CMMotionManager()
  private let queue = OperationQueue()

  private var accelBuffer: [(x: Double, y: Double, z: Double)] = []
  private var gyroBuffer: [(x: Double, y: Double, z: Double)] = []

  private var windowMs: Double = 20_000
  private var windowStartedAt: Date = Date()
  private var isSessionMode = false

  override static func requiresMainQueueSetup() -> Bool { true }

  override func supportedEvents() -> [String] {
    return ["onMotionWindow", "onMotionError"]
  }

  @objc(captureBaseline:resolver:rejecter:)
  func captureBaseline(durationMs: NSNumber,
                        resolver resolve: @escaping RCTPromiseResolveBlock,
                        rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard motionManager.isAccelerometerAvailable, motionManager.isGyroAvailable else {
      reject("SENSOR_UNAVAILABLE", "Accelerometer or gyroscope not available", nil)
      return
    }

    isSessionMode = false
    clearBuffers()
    startUpdates()

    let seconds = durationMs.doubleValue / 1000
    DispatchQueue.main.asyncAfter(deadline: .now() + seconds) { [weak self] in
      guard let self = self else { return }
      self.stopUpdates()
      resolve(self.featuresDictionary(self.computeFeatures()))
      self.clearBuffers()
    }
  }

  @objc(startSessionMonitoring:)
  func startSessionMonitoring(windowDurationMs: NSNumber) {
    guard motionManager.isAccelerometerAvailable, motionManager.isGyroAvailable else {
      sendEvent(withName: "onMotionError", body: ["code": "SENSOR_UNAVAILABLE"])
      return
    }
    isSessionMode = true
    windowMs = windowDurationMs.doubleValue
    windowStartedAt = Date()
    clearBuffers()
    startUpdates()
  }

  @objc func stopSessionMonitoring() {
    isSessionMode = false
    stopUpdates()
    clearBuffers()
  }

  private func startUpdates() {
    motionManager.accelerometerUpdateInterval = 1.0 / 50.0
    motionManager.gyroUpdateInterval = 1.0 / 50.0

    motionManager.startAccelerometerUpdates(to: queue) { [weak self] data, _ in
      guard let self = self, let data = data else { return }
      self.accelBuffer.append((data.acceleration.x, data.acceleration.y, data.acceleration.z))
      self.checkWindow()
    }
    motionManager.startGyroUpdates(to: queue) { [weak self] data, _ in
      guard let self = self, let data = data else { return }
      self.gyroBuffer.append((data.rotationRate.x, data.rotationRate.y, data.rotationRate.z))
    }
  }

  private func stopUpdates() {
    motionManager.stopAccelerometerUpdates()
    motionManager.stopGyroUpdates()
  }

  private func clearBuffers() {
    accelBuffer.removeAll()
    gyroBuffer.removeAll()
  }

  private func checkWindow() {
    guard isSessionMode else { return }
    if Date().timeIntervalSince(windowStartedAt) * 1000 >= windowMs {
      let features = computeFeatures()
      sendEvent(withName: "onMotionWindow", body: featuresDictionary(features))
      clearBuffers()
      windowStartedAt = Date()
    }
  }

  private struct MotionFeatures {
    let walkingStability: Double
    let stepRegularity: Double
    let stepIntervalVariability: Double
    let lateralSway: Double
    let motionVariance: Double
    let rotationVariance: Double
    let suddenMovement: Double
    let qualityScore: Int
  }

  private func computeFeatures() -> MotionFeatures {
    guard accelBuffer.count >= 10 else {
      return MotionFeatures(walkingStability: 0, stepRegularity: 0, stepIntervalVariability: 0,
                             lateralSway: 0, motionVariance: 0, rotationVariance: 0,
                             suddenMovement: 0, qualityScore: 20)
    }

    let gravity = 9.81
    let magnitudes = accelBuffer.map { sqrt($0.x * $0.x + $0.y * $0.y + $0.z * $0.z) * gravity }
    let linear = magnitudes.map { abs($0 - gravity) }

    let meanMag = linear.reduce(0, +) / Double(linear.count)
    let varianceMag = linear.map { ($0 - meanMag) * ($0 - meanMag) }.reduce(0, +) / Double(linear.count)
    let motionVariance = min(max(varianceMag, 0), 1)
    let walkingStability = min(max(1 - motionVariance, 0), 1)

    var stepIntervals: [Int] = []
    var lastPeakIndex = -1
    if linear.count > 2 {
      for i in 1..<(linear.count - 1) {
        let isPeak = linear[i] > meanMag && linear[i] >= linear[i - 1] && linear[i] >= linear[i + 1]
        if isPeak {
          if lastPeakIndex >= 0 { stepIntervals.append(i - lastPeakIndex) }
          lastPeakIndex = i
        }
      }
    }

    let stepRegularity: Double
    let stepIntervalVariability: Double
    if stepIntervals.count >= 2 {
      let meanInterval = Double(stepIntervals.reduce(0, +)) / Double(stepIntervals.count)
      let varInterval = stepIntervals.map { (Double($0) - meanInterval) * (Double($0) - meanInterval) }
        .reduce(0, +) / Double(stepIntervals.count)
      let stdDev = sqrt(varInterval)
      stepIntervalVariability = min(max(stdDev / (meanInterval + 1e-6), 0), 1)
      stepRegularity = min(max(1 - stepIntervalVariability, 0), 1)
    } else {
      stepRegularity = 0.5
      stepIntervalVariability = 0.5
    }

    let lateral = accelBuffer.map { sqrt($0.x * $0.x + $0.y * $0.y) * gravity }
    let lateralMean = lateral.reduce(0, +) / Double(lateral.count)
    let lateralSway = min(max(lateral.map { ($0 - lateralMean) * ($0 - lateralMean) }.reduce(0, +) / Double(lateral.count), 0), 1)

    let rotationMagnitudes = gyroBuffer.map { sqrt($0.x * $0.x + $0.y * $0.y + $0.z * $0.z) }
    let rotationVariance: Double
    if !rotationMagnitudes.isEmpty {
      let rotMean = rotationMagnitudes.reduce(0, +) / Double(rotationMagnitudes.count)
      rotationVariance = min(max(rotationMagnitudes.map { ($0 - rotMean) * ($0 - rotMean) }.reduce(0, +) / Double(rotationMagnitudes.count), 0), 1)
    } else {
      rotationVariance = 0
    }

    var suddenMovement = 0.0
    for i in 1..<linear.count {
      suddenMovement = max(suddenMovement, abs(linear[i] - linear[i - 1]))
    }

    let qualityScore: Int
    switch accelBuffer.count {
    case 500...: qualityScore = 95
    case 200..<500: qualityScore = 85
    case 50..<200: qualityScore = 70
    default: qualityScore = 50
    }

    return MotionFeatures(
      walkingStability: walkingStability, stepRegularity: stepRegularity,
      stepIntervalVariability: stepIntervalVariability, lateralSway: lateralSway,
      motionVariance: motionVariance, rotationVariance: rotationVariance,
      suddenMovement: min(max(suddenMovement, 0), 1), qualityScore: qualityScore,
    )
  }

  private func featuresDictionary(_ f: MotionFeatures) -> [String: Any] {
    return [
      "walkingStability": f.walkingStability,
      "stepRegularity": f.stepRegularity,
      "stepIntervalVariability": f.stepIntervalVariability,
      "lateralSway": f.lateralSway,
      "motionVariance": f.motionVariance,
      "rotationVariance": f.rotationVariance,
      "suddenMovement": f.suddenMovement,
      "qualityScore": f.qualityScore,
    ]
  }
}

File: android/app/src/main/java/com/soulalcohol/nativemonitoring/MotionModule.kt
package com.soulalcohol.nativemonitoring

import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Handler
import android.os.Looper
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Section 21/22: samples accelerometer + gyroscope into an in-memory
 * rolling buffer, extracts features (Section 23), then discards the raw
 * samples. Nothing here is ever persisted to SQLite (Section 39).
 *
 * NOT BUILD-VERIFIED: this Windows dev environment has no Android SDK, so
 * this file has not been compiled or run. Validate with a real
 * `./gradlew assembleDebug` / device run before shipping.
 */
class MotionModule(private val reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext), SensorEventListener {

  override fun getName() = "MotionModule"

  private val sensorManager: SensorManager by lazy {
    reactContext.getSystemService(android.content.Context.SENSOR_SERVICE) as SensorManager
  }
  private val accelerometer get() = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
  private val gyroscope get() = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)

  private val accelBuffer = mutableListOf<FloatArray>()
  private val gyroBuffer = mutableListOf<FloatArray>()

  private var windowMs: Long = 20_000
  private var windowStartedAt: Long = 0
  private var isSessionMode = false
  private var baselinePromise: Promise? = null
  private val mainHandler = Handler(Looper.getMainLooper())

  // --- Motion Baseline (Section 10): one-shot capture, resolves a Promise ---
  @ReactMethod
  fun captureBaseline(durationMs: Double, promise: Promise) {
    if (accelerometer == null || gyroscope == null) {
      promise.reject("SENSOR_UNAVAILABLE", "Accelerometer or gyroscope not available")
      return
    }
    isSessionMode = false
    baselinePromise = promise
    clearBuffers()
    registerListeners()

    // Handler on the main looper doesn't depend on currentActivity/decorView
    // being alive, so this still fires reliably if the app is backgrounded.
    mainHandler.postDelayed({
      unregisterListeners()
      val features = computeFeatures()
      baselinePromise?.resolve(featuresToMap(features))
      baselinePromise = null
      clearBuffers()
    }, durationMs.toLong())
  }

  // --- Continuous monitoring (Section 21/22): windowed feature events ---
  @ReactMethod
  fun startSessionMonitoring(windowDurationMs: Double) {
    if (accelerometer == null || gyroscope == null) {
      emitError("SENSOR_UNAVAILABLE")
      return
    }
    isSessionMode = true
    windowMs = windowDurationMs.toLong()
    windowStartedAt = System.currentTimeMillis()
    clearBuffers()
    registerListeners()
  }

  @ReactMethod
  fun stopSessionMonitoring() {
    isSessionMode = false
    unregisterListeners()
    clearBuffers()
  }

  @ReactMethod
  fun addListener(eventName: String) { /* required by RN event emitter contract */ }

  @ReactMethod
  fun removeListeners(count: Int) { /* required by RN event emitter contract */ }

  private fun registerListeners() {
    sensorManager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_GAME)
    sensorManager.registerListener(this, gyroscope, SensorManager.SENSOR_DELAY_GAME)
  }

  private fun unregisterListeners() {
    sensorManager.unregisterListener(this)
  }

  private fun clearBuffers() {
    accelBuffer.clear()
    gyroBuffer.clear()
  }

  override fun onSensorChanged(event: SensorEvent) {
    when (event.sensor.type) {
      Sensor.TYPE_ACCELEROMETER -> accelBuffer.add(event.values.copyOf())
      Sensor.TYPE_GYROSCOPE -> gyroBuffer.add(event.values.copyOf())
    }

    if (isSessionMode && System.currentTimeMillis() - windowStartedAt >= windowMs) {
      val features = computeFeatures()
      emitEvent("onMotionWindow", featuresToMap(features))
      clearBuffers()
      windowStartedAt = System.currentTimeMillis()
    }
  }

  override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}

  private data class MotionFeatures(
    val walkingStability: Double,
    val stepRegularity: Double,
    val stepIntervalVariability: Double,
    val lateralSway: Double,
    val motionVariance: Double,
    val rotationVariance: Double,
    val suddenMovement: Double,
    val qualityScore: Int,
  )

  /** Section 23/24: if too few samples were collected, quality is marked low rather than guessing. */
  private fun computeFeatures(): MotionFeatures {
    if (accelBuffer.size < 10) {
      return MotionFeatures(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, qualityScore = 20)
    }

    val magnitudes = accelBuffer.map { v ->
      sqrt((v[0] * v[0] + v[1] * v[1] + v[2] * v[2]).toDouble())
    }
    val gravity = 9.81
    val linearMagnitudes = magnitudes.map { abs(it - gravity) }

    val meanMag = linearMagnitudes.average()
    val varianceMag = linearMagnitudes.map { (it - meanMag) * (it - meanMag) }.average()
    val motionVariance = varianceMag.coerceIn(0.0, 1.0)
    val walkingStability = (1.0 - motionVariance).coerceIn(0.0, 1.0)

    // Simple peak detection above the mean = one "step".
    val stepIntervals = mutableListOf<Int>()
    var lastPeakIndex = -1
    for (i in 1 until linearMagnitudes.size - 1) {
      val isPeak = linearMagnitudes[i] > meanMag &&
        linearMagnitudes[i] >= linearMagnitudes[i - 1] &&
        linearMagnitudes[i] >= linearMagnitudes[i + 1]
      if (isPeak) {
        if (lastPeakIndex >= 0) stepIntervals.add(i - lastPeakIndex)
        lastPeakIndex = i
      }
    }
    val stepRegularity: Double
    val stepIntervalVariability: Double
    if (stepIntervals.size >= 2) {
      val meanInterval = stepIntervals.average()
      val varianceInterval = stepIntervals.map { (it - meanInterval) * (it - meanInterval) }.average()
      val stdDevInterval = sqrt(varianceInterval)
      stepIntervalVariability = (stdDevInterval / (meanInterval + 1e-6)).coerceIn(0.0, 1.0)
      stepRegularity = (1.0 - stepIntervalVariability).coerceIn(0.0, 1.0)
    } else {
      stepRegularity = 0.5
      stepIntervalVariability = 0.5
    }

    val lateralValues = accelBuffer.map { v -> sqrt((v[0] * v[0] + v[1] * v[1]).toDouble()) }
    val lateralMean = lateralValues.average()
    val lateralSway = lateralValues.map { (it - lateralMean) * (it - lateralMean) }
      .average().coerceIn(0.0, 1.0)

    val rotationMagnitudes = gyroBuffer.map { v ->
      sqrt((v[0] * v[0] + v[1] * v[1] + v[2] * v[2]).toDouble())
    }
    val rotationVariance = if (rotationMagnitudes.isNotEmpty()) {
      val rotMean = rotationMagnitudes.average()
      rotationMagnitudes.map { (it - rotMean) * (it - rotMean) }.average().coerceIn(0.0, 1.0)
    } else 0.0

    val suddenMovement = linearMagnitudes.zipWithNext { a, b -> abs(b - a) }.maxOrNull() ?: 0.0

    val qualityScore = when {
      accelBuffer.size >= 500 -> 95
      accelBuffer.size >= 200 -> 85
      accelBuffer.size >= 50 -> 70
      else -> 50
    }

    return MotionFeatures(
      walkingStability, stepRegularity, stepIntervalVariability,
      lateralSway, motionVariance, rotationVariance, suddenMovement.coerceIn(0.0, 1.0), qualityScore,
    )
  }

  private fun featuresToMap(f: MotionFeatures): WritableMap {
    val map = Arguments.createMap()
    map.putDouble("walkingStability", f.walkingStability)
    map.putDouble("stepRegularity", f.stepRegularity)
    map.putDouble("stepIntervalVariability", f.stepIntervalVariability)
    map.putDouble("lateralSway", f.lateralSway)
    map.putDouble("motionVariance", f.motionVariance)
    map.putDouble("rotationVariance", f.rotationVariance)
    map.putDouble("suddenMovement", f.suddenMovement)
    map.putInt("qualityScore", f.qualityScore)
    return map
  }

  private fun emitEvent(name: String, data: WritableMap) {
    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(name, data)
  }

  private fun emitError(code: String) {
    val map = Arguments.createMap()
    map.putString("code", code)
    emitEvent("onMotionError", map)
  }
}
